### ISSUE TITLE
[SUGGESTION] Mandating use of concepts with templates and compile-checking of templates against concepts

### ISSUE THREAD

**Issue by equeim** [hsutter/cppfront#253]
What do you think of adding requirement to specify concepts for all template parameters, and also checking at compile time that all operations with template parameter types and their objects used in template must be declared in concepts (i.e. it would be compile error to use operator+ on object of template parameter type if its concept doesn't require operator+). I think it would help to improve quality of templated code. It would allow to discover badly-written templates that use types incorrectly immediately instead of relying on bug reports from users.

It would also eliminate notorious multiple-page-long compilation errors that force users to read internals of template functions by forcing library authors to define concepts correctly and therefore allowing compilers to provide concise compilation errors explaining what's wrong at the call site (instead of dumping entire call stack between the place where error happened and where user actually called a function).

Basically I propose that following C++ code will fail to compile when transformed to equivalent Cpp2 code:
```c++
// Compilation error in Cpp2, no constraints on template parameter
template<typename T>
bool are_equal(T&& t1, T&& t2) {
    return t1 == t2;
}

template<std::equality_comparable T>
bool are_equal(T&& t1, T&& t2) {
    // Compilation error in Cpp2, usage of operation not declared in constraints (>>!even if template is not instantiated!<<)
    [[maybe_unused]] auto sneaky_code = t1 + t2;

    return t1 == t2;
}
```

**Comment by jcanizales** [comment]
I like this very much. But how can we make it work, and at the same time allow the use of the STL, as it exists now, from Cpp2 code?

The only viable way I see is make Cpp2 code (so, new code) require typing of compile time parameters, the same way it does with runtime parameters. But calling old untyped code must still be allowed.

**Comment by DaemonSnake** [comment]
I'm completely opposed to that.

If you like generics, that's fine,  google's Carbon is there for you.
That google wants to put its c++ devs into prison because they work with a mono-repo, making, in a sense, any of their c++ devs into a library vendor, is their cross to bear not ours.

Implementing this turns templates into a shore where each time you want to write some template or update it, you will be forced to write and modify the associated concept.
This leads to concept inflation and boilerplate and offers no value to the devs.
Also, as auto-lambda are unconstrained templates, this would force the declaration of global concepts each time you want to use an auto-lambda that does more than a single operation, or a lambda with N constraints via it's requires clause.

I believe what you'd really want is 'auto-concept'
ie: when writing a template the compiler generates automatically the associated concept for better diagnostics.
This way library vendors can specify explicitly concepts for better readability and those that don't bother will both make the template user benefit from it regardless.

Or maybe just better template instantiation errors.

**Comment by jcanizales** [comment]
C++ (the runtime language) is statically typed and has automatic type deduction; both things aren't mutually exclusive.

In my mind the crux of the question is, is duck typing the best for the compile-time language? The standard is slowly moving the compile-time language from purely functional duck-typed, towards being literally the same as the runtime language. And Herb's metaclass proposals move further in that direction.

Whether that makes some task better or worse, IMO, has to be discerned with specific examples that test specific proposals. Otherwise it's just philosophy.

**Comment by hsutter** [comment]
Thanks! Concepts are great, but I don't think we're ready to require them universally... there are lot of useful unconstrained templates.
